rfml.go

```diff
@@ -22,27 +22,122 @@ type fileParseError struct {
 	parseError error
 }
 
+// parsedTest is a convenience type that includes a path to the RFML file and
+// the parsed contents.
+type parsedTest struct {
+	filePath string
+	content  *rainforest.RFTest
+}
+
 func (e fileParseError) Error() string {
 	return fmt.Sprintf("%v:%v", e.filePath, e.parseError.Error())
 }
 
 // validateRFML is a wrapper around two other validation functions
 // first one for the single file and the other for whole directory
-func validateRFML(c cliContext) error {
+func validateRFML(c cliContext, api rfmlAPI) error {
 	if path := c.Args().First(); path != "" {
 		err := validateSingleRFMLFile(path)
 		if err != nil {
 			return cli.NewExitError(err.Error(), 1)
 		}
 		return nil
 	}
-	err := validateRFMLFilesInDirectory(c.String("test-folder"))
+	tests, err := readRFMLFiles([]string{c.String("test-folder")}, nil)
+	if err != nil {
+		return cli.NewExitError(err.Error(), 1)
+	}
+	err = validateRFMLFiles(tests, api, false)
 	if err != nil {
 		return cli.NewExitError(err.Error(), 1)
 	}
 	return nil
 }
 
+// readRFMLFiles takes in a list of files and/or directories and a list of tags
+// and returns a list of the parsed tests, or an error if it is encountered. To
+// allow all tags, pass in nil for tags.
+func readRFMLFiles(files []string, tags []string) ([]parsedTest, error) {
+	fileList := []string{}
+	for _, file := range files {
+		stat, err := os.Stat(file)
+		if err != nil {
+			return nil, err
+		}
+		if !stat.IsDir() {
+			if strings.HasSuffix(file, ".rfml") {
+				fileList = append(fileList, file)
+				continue
+			} else {
+				log.Printf("%s is not a valid RFML file", file)
+				continue
+			}
+		}
+
+		// We have a directory, walk through and find RFML files
+		err = filepath.Walk(file, func(path string, f os.FileInfo, err error) error {
+			if strings.HasSuffix(path, ".rfml") {
+				fileList = append(fileList, path)
+			}
+			return nil
+		})
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	parsedTests := []parsedTest{}
+	seenPaths := map[string]bool{}
+	for _, filePath := range fileList {
+		// No dups!
+		if seenPaths[filePath] {
+			continue
+		}
+		seenPaths[filePath] = true
+		test, err := readRFMLFile(filePath)
+		if err != nil {
+			return nil, err
+		}
+		if tags == nil || anyMember(test.Tags, tags) {
+			parsedTests = append(parsedTests, parsedTest{filePath, test})
```

Comment:
Yeah, I think it would be cleaner if we didn't have duplicate fields on different levels of structs.